File: ParalellAndVectorizedMatrices/src/main/java/org/example/ParallelMatrixMultiplicationOpenMP.java
package org.example;

import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

public class ParallelMatrixMultiplicationOpenMP implements ParallelMatrixMultiplication {

    @Override
    public double[][] multiply(double[][] A, double[][] B, int threads) throws InterruptedException {
        int size = A.length;
        int threshold = calculateThreshold(size);

        double[][] C = new double[size][size];

        ForkJoinPool pool = new ForkJoinPool(threads);
        pool.invoke(new MatrixMultiplyTask(A, B, C, 0, size, threshold));

        return C;
    }

    private static int calculateThreshold(int size) {
        return size / 10;
    }

    private static class MatrixMultiplyTask extends RecursiveAction {
        private final double[][] A, B, C;
        private final int startRow, endRow, threshold;

        public MatrixMultiplyTask(double[][] A, double[][] B, double[][] C, int startRow, int endRow, int threshold) {
            this.A = A;
            this.B = B;
            this.C = C;
            this.startRow = startRow;
            this.endRow = endRow;
            this.threshold = threshold;
        }

        @Override
        protected void compute() {
            int rows = endRow - startRow;
            if (rows <= threshold) {
                for (int i = startRow; i < endRow; i++) {
                    for (int j = 0; j < C[0].length; j++) {
                        double sum = 0;
                        for (int k = 0; k < B.length; k++) {
                            sum += A[i][k] * B[k][j];
                        }
                        C[i][j] = sum;
                    }
                }
            } else {
                int midRow = startRow + rows / 2;
                invokeAll(
                        new MatrixMultiplyTask(A, B, C, startRow, midRow, threshold),
                        new MatrixMultiplyTask(A, B, C, midRow, endRow, threshold)
                );
            }
        }
    }
}

File: ParallelAndVectorized/Algorithms/ParallelMatrixMultiplicationOpenMP.py
import numpy as np
import concurrent.futures


class ParallelMatrixMultiplicationOpenMP:

    def multiply(self, A, B, threads):
        size = A.shape[0]
        threshold = self.calculate_threshold(size)

        C = np.zeros((size, size), dtype=np.float32)

        with concurrent.futures.ThreadPoolExecutor(max_workers=threads) as executor:
            futures = [executor.submit(self._multiply_task, A, B, C, 0, size, threshold)]
            for future in concurrent.futures.as_completed(futures):
                future.result()

        return C

    def _multiply_task(self, A, B, C, start_row, end_row, threshold):
        rows = end_row - start_row
        if rows <= threshold:
            for i in range(start_row, end_row):
                for j in range(B.shape[1]):
                    sum_ = 0
                    for k in range(A.shape[1]):
                        sum_ += A[i, k] * B[k, j]
                    C[i, j] = sum_
        else:
            mid_row = start_row + rows // 2
            futures = []
            with concurrent.futures.ThreadPoolExecutor() as executor:
                futures.append(executor.submit(self._multiply_task, A, B, C, start_row, mid_row, threshold))
                futures.append(executor.submit(self._multiply_task, A, B, C, mid_row, end_row, threshold))
                for future in concurrent.futures.as_completed(futures):
                    future.result()

    def calculate_threshold(self, size):
        return size // 10
